itests/conftest.py

```diff
@@ -0,0 +1,187 @@
+    path = _expanduser(
+        glbl_cfg().get_host_item('run directory')
+    )
+    path.mkdir(exist_ok=True)
+    yield path
+
+
+@pytest.fixture(scope='session')
+def ses_test_dir(request, run_dir):
+    """The root reg dir for test flows in this test session."""
+    timestamp = get_current_time_string(use_basic_format=True)
+    uuid = f'cit-{timestamp}'
+    path = Path(run_dir, uuid)
+    path.mkdir(exist_ok=True)
+    yield path
+    _rm_if_empty(path)
+
+
+@pytest.fixture(scope='module')
+def mod_test_dir(request, ses_test_dir):
+    """The root reg dir for test flows in this test module."""
+    path = Path(ses_test_dir, request.module.__name__)
+    path.mkdir(exist_ok=True)
+    yield path
+    _rm_if_empty(path)
+
+
+@pytest.fixture
+def test_dir(request, mod_test_dir):
+    """The root reg dir for test flows in this test function."""
+    path = Path(mod_test_dir, request.function.__name__)
+    path.mkdir(parents=True, exist_ok=True)
+    yield path
+    _rm_if_empty(path)
+
+
+@pytest.fixture(scope='session')
+def ses_make_flow(run_dir, ses_test_dir):
+    """A function for creating session-level flows."""
+    yield partial(_make_flow, run_dir, ses_test_dir)
+
+
+@pytest.fixture(scope='module')
+def mod_make_flow(run_dir, mod_test_dir):
+    """A function for creating module-level flows."""
+    yield partial(_make_flow, run_dir, mod_test_dir)
+
+
+@pytest.fixture
+def make_flow(run_dir, test_dir):
+    """A function for creating function-level flows."""
+    yield partial(_make_flow, run_dir, test_dir)
+
+
+@pytest.fixture(scope='session')
+def ses_make_scheduler():
+    """Return a scheduler object for a flow."""
+    return _make_scheduler
+
+
+@pytest.fixture(scope='module')
+def mod_make_scheduler():
+    """Return a scheduler object for a flow."""
+    return _make_scheduler
+
+
+@pytest.fixture
+def make_scheduler():
+    """Return a scheduler object for a flow."""
+    return _make_scheduler
+
+
+@pytest.fixture(scope='session')
+def ses_flow(ses_make_flow, ses_make_scheduler):
+    """Make a session-level flow and return a scheduler object."""
+    return partial(_flow, ses_make_flow, ses_make_scheduler)
+
+
+@pytest.fixture(scope='module')
+def mod_flow(mod_make_flow, mod_make_scheduler):
+    """Make a module-level flow and return a scheduler object."""
+    return partial(_flow, mod_make_flow, mod_make_scheduler)
+
+
+@pytest.fixture
+def flow(make_flow, make_scheduler):
+    """Make a function-level flow and return a scheduler object."""
+    return partial(_flow, make_flow, make_scheduler)
+
+
+@pytest.fixture(scope='session')
+def ses_run_flow(run_dir, caplog):
+    """Run a session-level flow."""
+    caplog.set_level(logging.DEBUG, logger=CYLC_LOG)
+    return partial(_run_flow, run_dir, caplog)
+
+
+@pytest.fixture(scope='module')
+def mod_run_flow(run_dir):
+    """Run a module-level flow."""
+    return partial(_run_flow, run_dir, None)
+
+
+@pytest.fixture
+def run_flow(run_dir):
```

Comment:
The basic fixtures are:

* `make_flow` - writes the `suite.rc` to disk.
* `make_scheduler` - initiates the `Scheduler`.
* `flow` - shorthand for `make_scheduler(make_flow)` because I'm lazy.
* `run_flow` - calls `Scheduler.start`.

There are `mod_*` and `ses_*` variants of each of these to allow you to create Pytest fixtures with module or session level scoping. This allows us to run a workflow once and use it in multiple tests for efficiency reasons.

This is parallel safe!

Pytest has been configured to run tests from the same module together, so all the tests in a module can share the same scheduler, no problem.

*caveat* the exception is you can't run workflows in the session scope, but that would be crazy anyway.